### PR TITLE
feat: implement CreateDeal keeper, Tx and Query

### LINKED ISSUE #409: Implement a CreateDeal transaction & CLI
Implement a `CreateDeal` transaction & client cli for buyer to make a new `Deal`
- add `CreateDeal` keeper function that set initial state, make new account of `dealAddress` and send coins
- add `cmdCreateDeal()` client cli that create new `Deal` & broadcast transaction

### PR DESCRIPTION
**Pull request by audtlr24**
Closes #409

- I tested `CmdCreateDeal()`, `CmdGetDeals()` and `CmdGetDeal()` manually and works well

### CODE REVIEW

**Review by gyuguen — COMMENTED** [review]
Thank you!
I left some comment.

BTW, There's too much stuff in this PR.
`CreateDeal, GetDeal, ImportGenesis, ExportGenesis...`
I think it would be good to divide it into several PRs.

File: x/datadeal/keeper/deal.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		return types.Deal{}, err
+	}
+
+	return deal, nil
+}
+
+func (k Keeper) SetDeal(ctx sdk.Context, deal types.Deal) {
+	store := ctx.KVStore(k.storeKey)
+	dealKey := types.GetKeyPrefixDeals(deal.GetId())
+	bz := k.cdc.MustMarshalLengthPrefixed(&deal)
```
Comment (inchori) [inline x/datadeal/keeper/deal.go:109]: I think that it changes to `MarshalLengthPrefixed`.

**Review by inchori — COMMENTED** [review]
Thanks for your hard work!
I left a few comments.

File: x/datadeal/keeper/deal.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	return dealNumber, nil
+}
+
+func (k Keeper) GetDeal(ctx sdk.Context, dealID uint64) (types.Deal, error) {
+	store := ctx.KVStore(k.storeKey)
+	dealKey := types.GetKeyPrefixDeals(dealID)
+	if !store.Has(dealKey) {
+		return types.Deal{}, sdkerrors.Wrapf(types.ErrDealNotFound, "deal with ID %d does not exist", dealID)
+	}
+	bz := store.Get(dealKey)
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/keeper/deal.go]: Thanks, I fixed it in [a5bf7ea](https://github.com/medibloc/panacea-core/pull/423/commits/a5bf7eace5a66d3633fdee53a5e981d6affd09a9).

File: x/datadeal/genesis.go [hunk 3]
```diff
@@ -9,10 +9,22 @@ import (
...
 // ExportGenesis returns the capability module's exported genesis.
 func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
-	panic("implements me")
+	nextDealNum, err := k.GetNextDealNumber(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	return &types.GenesisState{
+		Deals:          nil,
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/genesis.go]: Thanks, I fixed it in [a5bf7ea](https://github.com/medibloc/panacea-core/pull/423/commits/a5bf7eace5a66d3633fdee53a5e981d6affd09a9). And I add test code in [28def23](https://github.com/medibloc/panacea-core/pull/423/commits/28def236c7f69f163908028951df1e3db20b09c0).

File: x/datadeal/genesis.go [hunk 4]
```diff
@@ -9,10 +9,22 @@ import (
...
+		NextDealNumber: nextDealNum,
+	}
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/genesis.go:35]: Thanks, I add it in [28def23](https://github.com/medibloc/panacea-core/pull/423/commits/28def236c7f69f163908028951df1e3db20b09c0)

File: x/datadeal/keeper/deal.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	bz := store.Get(types.KeyDealNextNumber)
+
+	val := gogotypes.UInt64Value{}
+
+	err := k.cdc.UnmarshalLengthPrefixed(bz, &val)
+	if err != nil {
+		return 0, err
+	}
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/keeper/deal.go]: Thanks, I fix all of these error handling code style to you recommend.

File: x/datadeal/genesis_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	suite.Require().NoError(err)
+
+	_, err = suite.DataDealKeeper.CreateDeal(suite.Ctx, acc1, deal2)
+	suite.Require().NoError(err)
+
+	genesisStatus := datadeal.ExportGenesis(suite.Ctx, suite.DataDealKeeper)
+	suite.Require().Equal(deal1.Address, genesisStatus.Deals[0].Address)
+	suite.Require().Equal(deal2.Address, genesisStatus.Deals[1].Address)
+	suite.Require().Equal(uint64(3), genesisStatus.NextDealNumber)
+}
```
Comment (inchori) [inline x/datadeal/genesis_test.go:83]: I think you should write comparing the all of the fields of deal.

File: x/datadeal/types/errors.go [hunk 7]
```diff
@@ -1,6 +1,13 @@
 package types
 
+import sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/types/errors.go:3]: Thanks, I talk with @inchori and that PR is merged and solve the conflicts after this PR merge.

File: x/datadeal/genesis_test.go (hunk 6 above)
Comment (audtlr24) [reply x/datadeal/genesis_test.go]: Okay, I fix it. Thanks!

File: x/datadeal/keeper/deal_test.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func TestDealTestSuite(t *testing.T) {
+	suite.Run(t, new(dealTestSuite))
+}
+
+var (
+	acc1         = sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address())
+	defaultFunds = sdk.NewCoins(sdk.NewCoin(assets.MicroMedDenom, sdk.NewInt(10000000000)))
+)
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/keeper/deal_test.go]: Thanks, I fix it in [f3e4a22](https://github.com/medibloc/panacea-core/pull/423/commits/f3e4a229be9939f94595185a907e25a09ec2ee2c). And I add `DataDealBaseTestSuite` that you recommend.

File: x/datadeal/keeper/deal.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	acc := k.accountKeeper.GetAccount(ctx, dealAddress)
+	if acc != nil {
+		return 0, sdkerrors.Wrapf(types.ErrDealAlreadyExist, "deal %d already exist", dealID)
+	}
+
+	acc = k.accountKeeper.NewAccount(ctx, authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(
+			dealAddress,
+		),
+		newDeal.GetAddress()),
```
Comment (H4NLee) [inline x/datadeal/keeper/deal.go]: This should be the name of module, which means `deal{dealID}`, not the deal address.

File: x/datadeal/keeper/deal_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (inchori) [reply x/datadeal/keeper/deal_test.go]: The test code of sell data will use `DataDealBaseTestSuite`, too.

File: x/datadeal/types/deal.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+		Status:       DEAL_STATUS_ACTIVE,
+	}
+}
+
+func NewDealAddress(dealID uint64) sdk.AccAddress {
+	dealKey := "deal" + strconv.FormatUint(dealID, 10)
+	return authtypes.NewModuleAddress(dealKey)
+}
+
+func (m Deal) ValidateBasic() error {
```
Comment (H4NLee) [inline x/datadeal/types/deal.go:32]: Let's validate `CurNumData` as well :)

**Review by H4NLee — COMMENTED** [review]
As @gyuguen said, I agree to make PR much smaller.
Left few comments.

File: x/datadeal/keeper/deal.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+	if !store.Has(types.KeyDealNextNumber) {
+		return 0, types.ErrDealNotInitialized
+	}
+
+	bz := store.Get(types.KeyDealNextNumber)
+
+	val := gogotypes.UInt64Value{}
+
+	if err := k.cdc.UnmarshalLengthPrefixed(bz, &val); err != nil {
+		return 0, err
```
Comment (gyuguen) [inline x/datadeal/keeper/deal.go:77]: It seems good to wrap errors with sdkerrors.

**Review by gyuguen — APPROVED** [review]
LGTM 👍

File: x/datadeal/types/keys.go [hunk 12]
```diff
@@ -22,10 +22,20 @@ const (
...
 )
 
 // CombineKeys function defines combines deal_id with data_hash.
-func CombineKeys(keys ...[]byte) []byte {
-	return bytes.Join(keys, KeyIndexSeparator)
+//func CombineKeys(keys ...[]byte) []byte {
+//	return bytes.Join(keys, KeyIndexSeparator)
+//}
+
+func GetKeyPrefixDeals(dealID uint64) []byte {
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/types/keys.go]: Thanks, I fix it.

File: x/datadeal/types/expected_keepers.go [hunk 13]
```diff
@@ -1 +1,28 @@
...
+	GetAccount(ctx sdk.Context, addr sdk.AccAddress) authtypes.AccountI
+
+	ValidatePermissions(macc authtypes.ModuleAccountI) error
+
+	GetModuleAddress(moduleName string) sdk.AccAddress
+	GetModuleAddressAndPermissions(moduleName string) (addr sdk.AccAddress, permissions []string)
+	GetModuleAccountAndPermissions(ctx sdk.Context, moduleName string) (authtypes.ModuleAccountI, []string)
+	GetModuleAccount(ctx sdk.Context, moduleName string) authtypes.ModuleAccountI
+	SetModuleAccount(ctx sdk.Context, macc authtypes.ModuleAccountI)
+}
```
[the remark this answers is not in the document]
Comment (audtlr24) [reply x/datadeal/types/expected_keepers.go]: Thanks, I delete unused functions 👍

File: x/datadeal/types/deal.go (hunk 10 above)
Comment (audtlr24) [reply x/datadeal/types/deal.go:32]: Thanks, I add it

File: x/datadeal/keeper/deal.go (hunk 9 above)
Comment (audtlr24) [reply x/datadeal/keeper/deal.go]: Thanks, I convert it to `strconv.FormatUint(newDeal.Id, 10))`. But, I'm not sure if it's better than address. Can you please explain?

**Review by inchori — APPROVED** [review]
LGTM 👍

File: x/datadeal/keeper/deal.go (hunk 11 above)
Comment (audtlr24) [reply x/datadeal/keeper/deal.go]: Thanks, I add sdkerrors.

File: x/datadeal/keeper/deal.go (hunk 9 above)
Comment (H4NLee) [reply x/datadeal/keeper/deal.go]: When you see `authtypes.NewModuleAccount()`, it accepts the name of module as a second argument.
So in order to make valid module account, you may need to pass the "deal{dealID}" (e.g. "deal1").
You can check the validity of module account as the below test code :)

```go
func (suite dealTestSuite) TestModuleAccount() {
	budget := &sdk.Coin{Denom: assets.MicroMedDenom, Amount: sdk.NewInt(10000000)}

	tempDeal := types.Deal{
		DataSchema:   []string{"http://jsonld.com"},
		Budget:       budget,
		MaxNumData:   10000,
		BuyerAddress: suite.acc1.String(),
	}

	newDeal := types.NewDeal(1, tempDeal)
	dealAddress, err := sdk.AccAddressFromBech32(newDeal.GetAddress())
	suite.Require().NoError(err)
	ma := authtypes.NewModuleAccount(authtypes.NewBaseAccountWithAddress(dealAddress), "deal1")
	suite.Require().NoError(ma.Validate())
}
```

File: x/datadeal/types/deal.go (hunk 10 above)
Comment (H4NLee) [reply x/datadeal/types/deal.go:32]: And also, it would be better to check the `CurNumData` is greater than 0 like `MaxNumData`

Comment (audtlr24) [reply x/datadeal/types/deal.go:32]: Thanks, I forgot. I will add it to next PR.

File: x/datadeal/keeper/deal.go (hunk 9 above)
Comment (audtlr24) [reply x/datadeal/keeper/deal.go]: Ah, Okay. I understand now. I have to use name when make `NewDealAddress`. I will reflect on the next PR.

**Pull request merged**